### ISSUE TITLE
Rebuild index

### ISSUE THREAD

**Issue by mnesrine** [restic/restic#1283]
## Output of `restic version`
restic 0.7.1
compiled with go1.8.3 on linux/amd64

## How did you run restic exactly?
export RESTIC_REPOSITORY=REPO
export AWS_ACCESS_KEY_ID=ACCESS_KEY
export AWS_SECRET_ACCESS_KEY=SECRET_ACCESS
export RESTIC_PASSWORD=PASSWORD
restic rebuild-index

## What backend/server/service did you use?

Minio

## Actual behavior
Error when we try to rebuild-index, we have about 250GiB of data.

```
counting files in repo
pack file cannot be listed be515cf2: Stat: Access Denied.
[1:05:04] 100.00%  49228 / 49228 packs
finding old index files
Put http://serverIP/1221-eyes/restic/index/f3bab2883871b0c3a6f9d7493af8206b62afa438dc09bf41b450e20bd3370524?partNumber=1&uploadId=4c7c5995-e549-4239-9d76-70918aca6f55: http: ContentLength=86604010 with Body length 84663141
client.PutObject
restic/backend/s3.(*Backend).Save
	src/restic/backend/s3/s3.go:228
restic/repository.(*Repository).SaveUnpacked
	src/restic/repository/repository.go:230
restic/repository.(*Repository).SaveJSONUnpacked
	src/restic/repository/repository.go:215
restic/index.Save
	src/restic/index/index.go:291
restic/index.(*Index).Save
	src/restic/index/index.go:262
main.rebuildIndex
	src/cmds/restic/cmd_rebuild_index.go:65
main.runRebuildIndex
	src/cmds/restic/cmd_rebuild_index.go:41
main.glob..func16
	src/cmds/restic/cmd_rebuild_index.go:19
github.com/spf13/cobra.(*Command).execute
	src/github.com/spf13/cobra/command.go:647
github.com/spf13/cobra.(*Command).ExecuteC
	src/github.com/spf13/cobra/command.go:726
github.com/spf13/cobra.(*Command).Execute
	src/github.com/spf13/cobra/command.go:685
main.main
	src/cmds/restic/main.go:63
runtime.main
	/usr/local/go/src/runtime/proc.go:185
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2197`
```

**Comment by fd0** [comment]
Hey, thank for leaving an issue. I can see two different things here:
 * restic was denied reading a file: `counting files in repo pack file cannot be listed be515cf2: Stat: Access Denied.`
 * A connection error occurred during upload: `Put http://serverIP/1221-eyes/restic/index/f3bab2883871b0c3a6f9d7493af8206b62afa438dc09bf41b450e20bd3370524?partNumber=1&uploadId=4c7c5995-e549-4239-9d76-70918aca6f55: http: ContentLength=86604010 with Body length 84663141`

I'd suggest two things: Upgrade to 0.7.3 (which includes a newer version of the minio-go library we're using to access minio) and re-run `prune`.

I'm going to close this issue for now, as it looks like it's not an issue in restic itself. Please feel free to add further comments and feedback how it goes!

**Comment by mnesrine** [comment]
Thanks to taking time to reply.
Finally in third try the rebuil-index has worked.
However I launched a prune on this bucket yesterday at 22H until I stopped it this morning. more than 8 hours and did not finish. what made it take too much time ?

Thank you.

**Comment by fd0** [comment]
Traversing all tree structures for the individual snapshots. We've just merged a PR (#1040) which adds a local cache that tremendously speeds up the `prune` operation. It is not contained in any released version of restic yet, but it's fully backwards compatible so you can easily just switch back to restic 0.7.3 or so.

If you want to give it a try, compile the code from the master branch, then run `prune` once (it will take as much time as before), then the next run of `prune` will be much faster (with the cache then).

**Comment by mnesrine** [comment]
Ok, I understand for the prune, now I upgrade to 0.7.3 version.
I have some questions, please could you clarify us: 
Does the prune rebuild-index ? 
And if not, when we should run 'rebuild-index' before or after 'prune' ?
In which case we must run 'check' ?

Thank you.

**Comment by zcalusic** [comment]
Yeah, prune runs rebuild-index **twice**! At the very beginning, and at the end of the operation. So, no need for extra rebuild-index if you run prune.

**Comment by fd0** [comment]
Indeed, `prune` already includes what `rebuild-index` does, so no need to run it separately.

It's a good idea to regularly run `restic check`, depending on your need maybe once a week or even after every backup. There's also `restic check --read-data` which will download **all** data from the repo and check it.

**Comment by mnesrine** [comment]
Indeed, prune take a lot of time (100 hours) but it finished successfully by releasing 70GiB over 260GiB of backups and has treated about 31800 packs.

However, I need some clarification to understand what happened, I done my prune with restic 0.7.3 (version with the cache) but when I relaunching the prune for the second time, he found less packs (11793) but it is still running since 16 hours,  why it is not faster ? Is there an instruction added to prune command to activate cache? and where I can found the local cache of prune ?

In my case I backup every day a big directory with not a lot of change on it, Is there a way to increase the size of blobs ? what do you think about this, can this make prune faster ?

Thank you.

**Comment by fd0** [comment]
> I done my prune with restic 0.7.3 (version with the cache)

I'm sorry, but the released version of 0.7.3 does not have the cache. The code for that is just in the master branch, not yet in any released version.

> when I relaunching the prune for the second time, he found less packs (11793) but it is still running since 16 hours, why it is not faster?

The prune code is rather dumb and will, when called manually, always run through the complete program. Apart from that without the exact output of restic and a description what you did in which order I cannot say more.

When the cache is available (you need to build restic yourself for that), prune will automatically use it.

> In my case I backup every day a big directory with not a lot of change on it, Is there a way to increase the size of blobs?

No, the blob size cannot be configured. And this also won't make prune faster.

I have several ideas on how to improve the speed of prune, but they're going to take time to implement.

**Comment by mnesrine** [comment]
Yeah, we test with a master branch.
Another question, can you confirm that the cache use by prune will be populated after a backup ?
and in this case we can not use the cache for our old backups ?
Thanks for reply.
